### PR TITLE
feat: support multiple recipients in rpc send transfer method, closes #5174

### LINKED ISSUE #5174: Multiple recipient bitcoin sends
Compliant with [WebBTC standard](https://wbips.netlify.app/wbips/WBIP005), Leather should support sending bitcoin to multiple recipients via its `sendTransfer` RPC API.

Any API changes **must** be backwards compatible.

### PR DESCRIPTION
**Pull request by alter-eggo**
> Try out Leather build 6fd3bc3 — [Extension build](https://github.com/leather-wallet/extension/actions/runs/8630590701), [Test report](https://leather-wallet.github.io/playwright-reports/feat/add-multi-recipient), [Storybook](https://feat-add-multi-recipient--65982789c7e2278518f189e7.chromatic.com), [Chromatic](https://www.chromatic.com/library?appId=65982789c7e2278518f189e7&branch=feat/add-multi-recipient)

This pr supports multiple recipients in rpc send transfer method

### CODE REVIEW

File: src/background/messaging/rpc-methods/send-transfer.ts [hunk 1]
```diff
@@ -43,8 +56,7 @@ export async function rpcSendTransfer(message: SendTransferRequest, port: chrome
   }
 
   const requestParams: RequestParams = [
-    ['address', message.params.address],
-    ['amount', message.params.amount],
+    ['recipients', JSON.stringify(params.recipients)],
```
Comment (fbwoolf) [inline src/background/messaging/rpc-methods/send-transfer.ts]: I think we will have to handle being backward compatible here so we don't break anyone using the old api?

File: src/app/pages/rpc-send-transfer/use-rpc-send-transfer.ts [hunk 2]
```diff
@@ -13,17 +14,18 @@ export function useRpcSendTransferRequestParams() {
   return useMemo(
     () => ({
       ...defaultParams,
-      address: initialSearchParams.get('address') ?? '',
-      amount: initialSearchParams.get('amount') ?? '',
       requestId: initialSearchParams.get('requestId') ?? '',
+      recipients: JSON.parse(
+        initialSearchParams.get('recipients') ?? '[]'
+      ) as RpcSendTransferRecipient[],
```
Comment (kyranjamie) [inline src/app/pages/rpc-send-transfer/use-rpc-send-transfer.ts]: Dangerous operation, can throw.

Query Params have their own way of handling arrays, by repeating the identifer

File: src/app/pages/rpc-send-transfer/use-rpc-send-transfer.ts [hunk 3]
```diff
@@ -32,13 +34,14 @@ export function useRpcSendTransfer() {
   if (!origin) throw new Error('Invalid params');
 
   return {
-    address,
-    amount,
+    recipients,
     origin,
     utxos,
+    totalAmount: recipients.reduce((acc, { amount }) => acc + Number(amount), 0),
```
Comment (kyranjamie) [inline src/app/pages/rpc-send-transfer/use-rpc-send-transfer.ts]:
```suggestion
    totalAmount: sumNumbers(recipients.map(({ amount }) => amount))
```

uses helper, avoids floating point maths

File: src/background/messaging/rpc-methods/send-transfer.ts [hunk 4]
```diff
@@ -16,7 +20,12 @@ import {
   triggerRequestWindowOpen,
 } from '../messaging-utils';
 
-export async function rpcSendTransfer(message: SendTransferRequest, port: chrome.runtime.Port) {
+export async function rpcSendTransfer(
+  message: RpcRequest<'sendTransfer', RpcSendTransferParams | SendTransferRequestParams>,
+  port: chrome.runtime.Port
+) {
+  let params = message.params as RpcSendTransferParams;
```
Comment (kyranjamie) [inline src/background/messaging/rpc-methods/send-transfer.ts]: Imperative coding style. How can we do without reassigning var?

File: src/background/messaging/rpc-methods/send-transfer.ts (hunk 1 above)
Comment (alter-eggo) [reply src/background/messaging/rpc-methods/send-transfer.ts]: I convert old params to new ones above

File: src/app/pages/rpc-send-transfer/use-rpc-send-transfer.ts [hunk 5]
```diff
@@ -32,13 +33,17 @@ export function useRpcSendTransfer() {
...
   return {
-    address,
-    amount,
+    recipients: recipients.map((address, index) => ({
+      address,
+      amount: amounts[index],
+    })),
     origin,
     utxos,
+    totalAmount: sumNumbers(amounts.map(Number)),
```
Comment (kyranjamie) [inline src/app/pages/rpc-send-transfer/use-rpc-send-transfer.ts:42]: What does passing number as the callback do 🤔

File: src/background/messaging/rpc-methods/send-transfer.ts (hunk 4 above)
Comment (alter-eggo) [reply src/background/messaging/rpc-methods/send-transfer.ts]: refactored a bit
but left variable, we need to assign somewhere converted params 🤔

File: src/app/pages/rpc-send-transfer/use-rpc-send-transfer.ts (hunk 5 above)
Comment (alter-eggo) [reply src/app/pages/rpc-send-transfer/use-rpc-send-transfer.ts:42]: converts to num (copilot help)

**Comment by markmhendrickson** [comment]
@alter-eggo is there a way to test this in the test app or elsewhere?

**Comment by alter-eggo** [comment]
@markmhendrickson yes, I've added btn to test app
Screenshot 2024-04-09 at 15 43 40

**Comment by kyranjamie** [comment]
Screenshot 2024-04-09 at 13 55 36

**Review by fbwoolf — APPROVED** [review]

**Comment by alter-eggo** [comment]
@kyranjamie fixed, problem is on current dev as well

**Review by kyranjamie — APPROVED** [review]

**Comment by markmhendrickson** [comment]
@alter-eggo For updating the docs: It seems the developer can continue passing a string address value as `recipient` or an array of string values as `recipients`, correct?

**Comment by alter-eggo** [comment]
@markmhendrickson yes, it's backwards compatible

**Comment by markmhendrickson** [comment]
I'm seeing now that previously / currently we have `amount` as the parameter. So by backwards compatible, you mean both `amount` and `recipient` are still supported, correct? Do we have tests that enforce this?

**Comment by alter-eggo** [comment]
by backwards compatible I mean even if you pass `address` and `amount` - it will work. there are tests, but it can be tested as well in test app.
this fn converts legacy params to new ones:
```
export function convertRpcSendTransferLegacyParamsToNew(params: RpcSendTransferParamsLegacy) {
  return {
    recipients: [{ address: params.address, amount: params.amount }],
    network: params.network,
    account: params.account,
  };
}
```

**Pull request merged**